File: web/src/lib/cron.ts
// Client-side cron parser.
// Supports standard 5-field expressions: minute hour dom month dow
// and common aliases: @hourly, @daily, @weekly, @monthly, @yearly.

// Parsed representation of a single cron field (e.g. "1-5", "*/15", "1,3,5")
interface CronField {
  values: Set<number> // expanded set of matching values within [min, max]
}

export interface CronExpression {
  minute: CronField
  hour: CronField
  dom: CronField    // day of month
  month: CronField
  dow: CronField    // day of week (0=Sun, 6=Sat)
  raw: string       // original expression string
}

// Alias expansions to 5-field expressions
const ALIASES: Record<string, string> = {
  '@yearly':  '0 0 1 1 *',
  '@annually': '0 0 1 1 *',
  '@monthly': '0 0 1 * *',
  '@weekly':  '0 0 * * 0',
  '@daily':   '0 0 * * *',
  '@midnight': '0 0 * * *',
  '@hourly':  '0 * * * *',
}

// Field bounds: [min, max]
const FIELD_BOUNDS: [number, number][] = [
  [0, 59],  // minute
  [0, 23],  // hour
  [1, 31],  // dom
  [1, 12],  // month
  [0, 6],   // dow (0=Sun)
]

const DOW_NAMES: Record<string, number> = {
  sun: 0, mon: 1, tue: 2, wed: 3, thu: 4, fri: 5, sat: 6,
}

const MONTH_NAMES: Record<string, number> = {
  jan: 1, feb: 2, mar: 3, apr: 4, may: 5, jun: 6,
  jul: 7, aug: 8, sep: 9, oct: 10, nov: 11, dec: 12,
}

// Parse a single cron field token into a set of matching values
function parseField(token: string, min: number, max: number, names?: Record<string, number>): CronField | null {
  const values = new Set<number>()

  for (const part of token.split(',')) {
    const trimmed = part.trim()
    if (!trimmed) return null

    // Replace named values (e.g. MON, JAN)
    let resolved = trimmed.toLowerCase()
    if (names) {
      for (const [name, val] of Object.entries(names)) {
        resolved = resolved.replace(new RegExp(`\\b${name}\\b`, 'gi'), String(val))
      }
    }

    // Handle step: */2, 1-5/2, or just a number with step
    const stepMatch = resolved.match(/^(.+)\/(\d+)$/)
    const step = stepMatch ? parseInt(stepMatch[2], 10) : 1
    const range = stepMatch ? stepMatch[1] : resolved

    if (step < 1) return null

    if (range === '*') {
      for (let i = min; i <= max; i += step) values.add(i)
    } else if (range.includes('-')) {
      const [startStr, endStr] = range.split('-')
      const start = parseInt(startStr, 10)
      const end = parseInt(endStr, 10)
      if (isNaN(start) || isNaN(end) || start < min || end > max || start > end) return null
      for (let i = start; i <= end; i += step) values.add(i)
    } else {
      const val = parseInt(range, 10)
      if (isNaN(val) || val < min || val > max) return null
      values.add(val)
    }
  }

  return values.size > 0 ? { values } : null
}

// Parse a cron expression string into a CronExpression, or null if invalid
export function parseCron(expr: string): CronExpression | null {
  const trimmed = expr.trim()

  // Check for aliases first
  const aliased = ALIASES[trimmed.toLowerCase()]
  const effective = aliased ?? trimmed

  const parts = effective.split(/\s+/)
  if (parts.length !== 5) return null

  const namesByField: (Record<string, number> | undefined)[] = [
    undefined,    // minute: no names
    undefined,    // hour: no names
    undefined,    // dom: no names
    MONTH_NAMES,  // month
    DOW_NAMES,    // dow
  ]

  const fields: CronField[] = []
  for (let i = 0; i < 5; i++) {
    const field = parseField(parts[i], FIELD_BOUNDS[i][0], FIELD_BOUNDS[i][1], namesByField[i])
    if (!field) return null
    fields.push(field)
  }

  return {
    minute: fields[0],
    hour: fields[1],
    dom: fields[2],
    month: fields[3],
    dow: fields[4],
    raw: trimmed,
  }
}

// Compute the next time a cron expression matches after the given date.
// Returns null if the expression is invalid. Searches up to 2 years ahead.
export function nextCronRun(expr: string, after?: Date): Date | null {
  const cron = typeof expr === 'string' ? parseCron(expr) : null
  if (!cron) return null

  const start = after ? new Date(after) : new Date()
  // Advance to next minute boundary
  const candidate = new Date(start)
  candidate.setSeconds(0, 0)
  candidate.setMinutes(candidate.getMinutes() + 1)

  const limit = new Date(start)
  limit.setFullYear(limit.getFullYear() + 2)

  while (candidate < limit) {
    const month = candidate.getMonth() + 1  // JS months are 0-indexed
    const dom = candidate.getDate()
    const dow = candidate.getDay()
    const hour = candidate.getHours()
    const minute = candidate.getMinutes()

    if (
      cron.month.values.has(month) &&
      cron.dom.values.has(dom) &&
      cron.dow.values.has(dow) &&
      cron.hour.values.has(hour) &&
      cron.minute.values.has(minute)
    ) {
      return candidate
    }

    // Advance: skip by the smallest meaningful increment
    candidate.setMinutes(candidate.getMinutes() + 1)
  }

  return null
}

// Day-of-week names for human-readable descriptions
const DOW_LABELS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']
const MONTH_LABELS = ['', 'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December']

// Return a human-readable description of a cron expression.
// Returns empty string for invalid expressions.
export function describeCron(expr: string): string {
  const trimmed = expr.trim().toLowerCase()

  // Handle aliases with friendly names
  if (trimmed === '@hourly') return 'Every hour at :00'
  if (trimmed === '@daily' || trimmed === '@midnight') return 'Every day at midnight'
  if (trimmed === '@weekly') return 'Every Sunday at midnight'
  if (trimmed === '@monthly') return 'First day of every month at midnight'
  if (trimmed === '@yearly' || trimmed === '@annually') return 'January 1st at midnight'

  const cron = parseCron(expr)
  if (!cron) return ''

  const parts: string[] = []

  // Time portion
  const minutes = sorted(cron.minute.values)
  const hours = sorted(cron.hour.values)

  if (hours.length <= 3 && minutes.length <= 2) {
    // Enumerate specific times: "At 09:00, 17:00"
    const times = hours.flatMap(h =>
      minutes.map(m => `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`)
    )
    parts.push(`At ${times.join(', ')}`)
  } else {
    // Describe with intervals
    if (minutes.length === 60) {
      parts.push('Every minute')
    } else if (isStepPattern(minutes, 0)) {
      parts.push(`Every ${stepSize(minutes, 0)} minutes`)
    } else {
      parts.push(`At minute ${describeValues(minutes)}`)
    }

    if (hours.length < 24) {
      if (isStepPattern(hours, 0)) {
        parts.push(`every ${stepSize(hours, 0)} hours`)
      } else {
        parts.push(`past hour ${describeValues(hours)}`)
      }
    }
  }

  // Day-of-week constraint
  const dows = sorted(cron.dow.values)
  if (dows.length < 7) {
    if (dows.length === 5 && dows[0] === 1 && dows[4] === 5) {
      parts.push('on weekdays')
    } else if (dows.length === 2 && dows[0] === 0 && dows[1] === 6) {
      parts.push('on weekends')
    } else {
      parts.push(`on ${dows.map(d => DOW_LABELS[d]).join(', ')}`)
    }
  }

  // Day-of-month constraint
  const doms = sorted(cron.dom.values)
  if (doms.length < 31) {
    parts.push(`on day ${describeValues(doms)} of the month`)
  }

  // Month constraint
  const months = sorted(cron.month.values)
  if (months.length < 12) {
    parts.push(`in ${months.map(m => MONTH_LABELS[m]).join(', ')}`)
  }

  return parts.join(' ')
}

// Helpers for describeCron

function sorted(set: Set<number>): number[] {
  return Array.from(set).sort((a, b) => a - b)
}

function describeValues(values: number[]): string {
  if (values.length <= 5) return values.join(', ')
  return `${values[0]}-${values[values.length - 1]}`
}

function isStepPattern(values: number[], min: number): boolean {
  if (values.length < 2) return false
  const step = values[1] - values[0]
  if (step < 2) return false
  return values.every((v, i) => v === min + i * step)
}

function stepSize(values: number[], min: number): number {
  if (values.length < 2) return 1
  return values[1] - (values[0] === min ? min : values[0])
}

// Validate that a string is a valid cron expression (5-field or alias)
export function isValidCron(expr: string): boolean {
  return parseCron(expr) !== null
}

// Format a relative time from now to a future date, e.g. "in 2h 15m" or "Mon 9:00am"
export function formatRelativeTime(target: Date, now?: Date): string {
  const ref = now ?? new Date()
  const diffMs = target.getTime() - ref.getTime()
  if (diffMs < 0) return 'past'

  const diffMin = Math.floor(diffMs / 60_000)
  if (diffMin < 60) return `in ${diffMin}m`

  const diffHr = Math.floor(diffMin / 60)
  const remMin = diffMin % 60
  if (diffHr < 24) return remMin > 0 ? `in ${diffHr}h ${remMin}m` : `in ${diffHr}h`

  // More than 24h away — show day and time
  const dayName = DOW_LABELS[target.getDay()].slice(0, 3)
  const hour = target.getHours()
  const minute = target.getMinutes()
  const ampm = hour >= 12 ? 'pm' : 'am'
  const h12 = hour % 12 || 12
  const timeStr = minute > 0
    ? `${h12}:${String(minute).padStart(2, '0')}${ampm}`
    : `${h12}${ampm}`
  return `${dayName} ${timeStr}`
}

File: pkg/cron/cron.go
// Package cron provides a minimal, dependency-free parser for standard 5-field
// cron expressions (minute hour dom month dow) and common aliases.
//
// Supported syntax per field:
//   - Wildcards: *
//   - Ranges: 1-5
//   - Steps: */5, 1-10/2
//   - Lists: 1,3,5
//
// Aliases: @yearly, @annually, @monthly, @weekly, @daily, @midnight, @hourly
package cron

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Expression is a parsed cron expression that can compute the next matching time.
type Expression struct {
	minutes  []bool // [0..59]
	hours    []bool // [0..23]
	doms     []bool // [1..31]  (index 0 unused)
	months   []bool // [1..12]  (index 0 unused)
	dows     []bool // [0..6]   (0=Sunday)
	domSet   bool   // true when dom field is not wildcard
	dowSet   bool   // true when dow field is not wildcard
	original string
}

// String returns the original cron expression.
func (e *Expression) String() string { return e.original }

// aliases maps shorthand names to their 5-field equivalents.
var aliases = map[string]string{
	"@yearly":   "0 0 1 1 *",
	"@annually": "0 0 1 1 *",
	"@monthly":  "0 0 1 * *",
	"@weekly":   "0 0 * * 0",
	"@daily":    "0 0 * * *",
	"@midnight": "0 0 * * *",
	"@hourly":   "0 * * * *",
}

// Parse parses a 5-field cron expression or an alias and returns an Expression.
func Parse(expr string) (*Expression, error) {
	expr = strings.TrimSpace(expr)
	if expr == "" {
		return nil, fmt.Errorf("cron: empty expression")
	}

	// Resolve aliases.
	if strings.HasPrefix(expr, "@") {
		resolved, ok := aliases[strings.ToLower(expr)]
		if !ok {
			return nil, fmt.Errorf("cron: unknown alias %q", expr)
		}
		return Parse(resolved)
	}

	fields := strings.Fields(expr)
	if len(fields) != 5 {
		return nil, fmt.Errorf("cron: expected 5 fields, got %d in %q", len(fields), expr)
	}

	e := &Expression{
		minutes: make([]bool, 60),
		hours:   make([]bool, 24),
		doms:    make([]bool, 32), // index 0 unused; 1-31
		months:  make([]bool, 13), // index 0 unused; 1-12
		dows:    make([]bool, 7),
	}
	e.original = expr

	if err := parseField(fields[0], e.minutes, 0, 59); err != nil {
		return nil, fmt.Errorf("cron: minute field: %w", err)
	}
	if err := parseField(fields[1], e.hours, 0, 23); err != nil {
		return nil, fmt.Errorf("cron: hour field: %w", err)
	}
	if err := parseField(fields[2], e.doms, 1, 31); err != nil {
		return nil, fmt.Errorf("cron: day-of-month field: %w", err)
	}
	if err := parseField(fields[3], e.months, 1, 12); err != nil {
		return nil, fmt.Errorf("cron: month field: %w", err)
	}
	if err := parseDOWField(fields[4], e.dows); err != nil {
		return nil, fmt.Errorf("cron: day-of-week field: %w", err)
	}

	e.domSet = fields[2] != "*"
	e.dowSet = fields[4] != "*"

	return e, nil
}

// Next returns the next time at or after t that matches the expression.
// It searches up to 4 years ahead to handle leap year and rare patterns.
func (e *Expression) Next(t time.Time) (time.Time, error) {
	// Start from the next whole minute after t.
	t = t.Truncate(time.Minute).Add(time.Minute)

	// Search limit: 4 years to cover leap year edge cases.
	limit := t.Add(4 * 365 * 24 * time.Hour)

	for t.Before(limit) {
		// Check month.
		if !e.months[t.Month()] {
			// Advance to next month, day 1, 00:00.
			t = time.Date(t.Year(), t.Month()+1, 1, 0, 0, 0, 0, t.Location())
			continue
		}

		// Check day: when both dom and dow are restricted (not wildcard),
		// match if EITHER matches (standard cron union semantics).
		dayOK := e.matchDay(t)
		if !dayOK {
			t = time.Date(t.Year(), t.Month(), t.Day()+1, 0, 0, 0, 0, t.Location())
			continue
		}

		// Check hour.
		if !e.hours[t.Hour()] {
			t = time.Date(t.Year(), t.Month(), t.Day(), t.Hour()+1, 0, 0, 0, t.Location())
			continue
		}

		// Check minute.
		if !e.minutes[t.Minute()] {
			t = t.Add(time.Minute)
			continue
		}

		return t, nil
	}

	return time.Time{}, fmt.Errorf("cron: no matching time found within 4 years from %v", t)
}

// matchDay returns true if the given time matches the day-of-month/day-of-week
// constraints. Standard cron: if both dom and dow are restricted, match on either.
func (e *Expression) matchDay(t time.Time) bool {
	domOK := e.doms[t.Day()]
	dowOK := e.dows[int(t.Weekday())]

	if e.domSet && e.dowSet {
		// Both restricted: union semantics (match if either matches).
		return domOK || dowOK
	}
	// Only one restricted, or neither: both must match.
	return domOK && dowOK
}

// parseField parses a single cron field (minute, hour, dom, month) into a bool slice.
// lo and hi are the valid value range (inclusive).
func parseField(field string, bits []bool, lo, hi int) error {
	for _, part := range strings.Split(field, ",") {
		if err := parsePart(part, bits, lo, hi); err != nil {
			return err
		}
	}
	return nil
}

// parsePart parses one element of a comma-separated list: a value, range, wildcard,
// or any of these with a step suffix.
func parsePart(part string, bits []bool, lo, hi int) error {
	part = strings.TrimSpace(part)
	if part == "" {
		return fmt.Errorf("empty field part")
	}

	// Split on "/" for step.
	step := 1
	if idx := strings.IndexByte(part, '/'); idx >= 0 {
		s, err := strconv.Atoi(part[idx+1:])
		if err != nil || s <= 0 {
			return fmt.Errorf("invalid step %q", part[idx+1:])
		}
		step = s
		part = part[:idx]
	}

	// Determine range.
	var start, end int
	if part == "*" {
		start, end = lo, hi
	} else if idx := strings.IndexByte(part, '-'); idx >= 0 {
		var err error
		start, err = strconv.Atoi(part[:idx])
		if err != nil {
			return fmt.Errorf("invalid range start %q", part[:idx])
		}
		end, err = strconv.Atoi(part[idx+1:])
		if err != nil {
			return fmt.Errorf("invalid range end %q", part[idx+1:])
		}
	} else {
		v, err := strconv.Atoi(part)
		if err != nil {
			return fmt.Errorf("invalid value %q", part)
		}
		start, end = v, v
	}

	if start < lo || end > hi || start > end {
		return fmt.Errorf("value out of range [%d-%d]: %d-%d", lo, hi, start, end)
	}

	for v := start; v <= end; v += step {
		bits[v] = true
	}
	return nil
}

// parseDOWField is like parseField but handles 7 as an alias for 0 (Sunday).
func parseDOWField(field string, bits []bool) error {
	for _, part := range strings.Split(field, ",") {
		if err := parseDOWPart(part, bits); err != nil {
			return err
		}
	}
	return nil
}

func parseDOWPart(part string, bits []bool) error {
	part = strings.TrimSpace(part)
	if part == "" {
		return fmt.Errorf("empty field part")
	}

	step := 1
	if idx := strings.IndexByte(part, '/'); idx >= 0 {
		s, err := strconv.Atoi(part[idx+1:])
		if err != nil || s <= 0 {
			return fmt.Errorf("invalid step %q", part[idx+1:])
		}
		step = s
		part = part[:idx]
	}

	var start, end int
	if part == "*" {
		start, end = 0, 6
	} else if idx := strings.IndexByte(part, '-'); idx >= 0 {
		var err error
		start, err = strconv.Atoi(part[:idx])
		if err != nil {
			return fmt.Errorf("invalid range start %q", part[:idx])
		}
		end, err = strconv.Atoi(part[idx+1:])
		if err != nil {
			return fmt.Errorf("invalid range end %q", part[idx+1:])
		}
	} else {
		v, err := strconv.Atoi(part)
		if err != nil {
			return fmt.Errorf("invalid value %q", part)
		}
		start, end = v, v
	}

	// Normalize 7 → 0 (both mean Sunday).
	if start == 7 {
		start = 0
	}
	if end == 7 {
		end = 0
	}

	// Handle wrap-around ranges like 5-0 (Fri-Sun).
	if start > end {
		for v := start; v <= 6; v += step {
			bits[v] = true
		}
		for v := 0; v <= end; v += step {
			bits[v] = true
		}
		return nil
	}

	if start < 0 || end > 6 {
		return fmt.Errorf("day-of-week value out of range [0-7]: %d-%d", start, end)
	}

	for v := start; v <= end; v += step {
		bits[v] = true
	}
	return nil
}
